### PR TITLE
Implement webservice connector plugin

### PR DESCRIPTION
**Pull request by cameron1729**
# Testing
**NB**: Testing instructions assume a fresh install

1. Set up web services (ensure the role assigned to the web service user has the capability `tool/messagebroker:submitmessage`)
2. Make a curl request: `curl 'http://localhost/m/stable_401/webservice/rest/server.php?moodlewsrestformat=json' --data 'wsfunction=mbconnector_webservice_submit_message&topic=cool.movies&body=%7B%22some%22%3A%22thing%22%7D&wstoken=WS_TOKEN_HERE'`
3. **Verify** the response looks like `{"success":true,"stored":false,"receivers":0}`
4. Put the following script in to the root of any plugin and call it messagebroker.php:

```
<?php

class receiver implements tool_messagebroker\receiver\message_receiver {
    private string $topic;
    private string $style;
    private string $willsucceed;

    public function __construct(string $topic, string $style, bool $willsucceed) {
        $this->topic = $topic;
        $this->style = $style;
        $this->willsucceed = $willsucceed;
    }

    public static function instance(): self {
        return new self();
    }

    public function process_message(tool_messagebroker\message\immutable_message $message): bool {
        return $this->willsucceed;
    }

    public function get_preferred_message_processing_method(): string {
        return $this->style;
    }

    public function get_registered_topic(): string {
        return "/" . $this->topic . "/";
    }
}

function [FRANKENSTYLE PREFIX]_build_message_receivers() {
    return [
        new receiver('movies', tool_messagebroker\receiver\processing_style::EPHEMERAL, true),
        new receiver('movies', tool_messagebroker\receiver\processing_style::EPHEMERAL, true),
        new receiver('movies', tool_messagebroker\receiver\processing_style::EPHEMERAL, true)
    ];
}
```
5. Purge caches
6. Make another curl request (the same as before)
7. **Verify** the response is: `{"success":true,"stored":false,"receivers":3}`
8. Inspect the database and **verify** there are no records in the mdl_mbdatastore_standarddb_msg table
9. Alter the `build_message_receivers` function to look like:
```
function [FRANKENSTYLE PREFIX]_build_message_receivers() {
    return [
        new receiver('movies', tool_messagebroker\receiver\processing_style::EPHEMERAL, true),
        new receiver('movies', tool_messagebroker\receiver\processing_style::EPHEMERAL, true),
        new receiver('dvds', tool_messagebroker\receiver\processing_style::EPHEMERAL, true),
        new receiver('movies', tool_messagebroker\receiver\processing_style::EPHEMERAL, true)
    ];
}
```
10. Make the curl request again
11. **Verify** the response is: `{"success":true,"stored":false,"receivers":3}` and there are still no records in the DB
12. Alter the function to look like:
```
function [FRANKENSTYLE PREFIX]_build_message_receivers() {
    return [
        new receiver('movies', tool_messagebroker\receiver\processing_style::EPHEMERAL, true),
        new receiver('movies', tool_messagebroker\receiver\processing_style::EPHEMERAL, true),
        new receiver('dvds', tool_messagebroker\receiver\processing_style::EPHEMERAL, true),
        new receiver('movies', tool_messagebroker\receiver\processing_style::DURABLE, true)
    ];
}
```
13. Make the curl request and **verify** the response looks like: `{"success":true,"stored":true,"receivers":3}`
14. Check the DB and **verify** a record was created for the message
15. Browse to "Site Administration" > "Plugins" > "Admin tools" > "Message Broker"
16. Set "Message receiving mode" to "Ephemeral"
17. Make the curl request again and **verify** the response looks like: `{"success":true,"stored":false,"receivers":3}`
18. Inspect the DB and **verify** a record was **not** created for the message
19. Change the receiving mode to durable and remove the durable receiver from the receivers array
20. Make the curl request and **verify** the response looks like: `{"success":true,"stored":true,"receivers":2}`
21. Inspect the DB and **verify** a record was created for the message
22. Set the receiving mode back to receiver preference
23. Hack the file admin/tool/messagebroker/datastores/classes/durable_dao.php:
    1. Modify the `write_new_message` method so that it immediately throws an exception 
24. Add a durable receiver to the receivers array with the "movies" topic
25. Make the curl request and verify the response looks like: `{"success":false,"stored":false,"receivers":3}`
26. Inspect the DB and **verify** a record was **not** created for the message
27. Remove the `tool/messagebroker:submitmessage` capability from the web service user's role and attempt the curl request again
28. **Verify** it fails with the response: `{"exception":"required_capability_exception","errorcode":"nopermissions","message":"Sorry, but you do not currently have permissions to do that (Submit a message to the Message Broker)."}`
29. Play with various combinations of settings and receivers to verify everything works as expected

### CODE REVIEW

File: connectors/webservice/classes/external/receive_message.php [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+use external_value;
+use moodle_exception;
+use tool_messagebroker\message\received_message;
+use tool_messagebroker\message_processor;
+use tool_messagebroker\receiver\processing_style;
+
+class receive_message extends external_api {
+
+    public static function execute(string $topic, string $body) {
+        // TODO: authorisation
```
Comment (cameron1729) [inline connectors/webservice/classes/external/receive_message.php]: @darrencocco I'm not sure if anything is actually needed here? The service is behind a cap check - is that enough or do we need something else?

Comment (darrencocco) [reply connectors/webservice/classes/external/receive_message.php]: Last I checked the capabilities specified in services.php are just informational and don't actually get used to enforce authorisation checks.

File: classes/message_processor.php [hunk 2]
```diff
@@ -46,51 +48,55 @@ protected function __construct(string $receivermode, ?durable_dao_interface $dao
...
         }
     }
 
-    /**
-     * @param immutable_message $message
-     * @param message_receiver[] $receivers
-     * @return void
-     */
-    protected function process_new_message(immutable_message $message, array $receivers) {
+    protected function process_new_message(immutable_message $message, array $receivers): array {
```
Comment (darrencocco) [inline classes/message_processor.php:71]: Neat rewrite.

File: connectors/webservice/classes/external/receive_message.php (hunk 1 above)
Comment (cameron1729) [reply connectors/webservice/classes/external/receive_message.php]: Wow, TIL. It's [even documented](https://moodledev.io/docs/apis/subsystems/external/writing-a-service)!

image

Pushed a commit to add an explicit cap check and updated testing instructions.

**Comment by darrencocco** [comment]
Looks good

**Pull request merged**